bpf/lib/policy.h

```diff
@@ -48,49 +49,56 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
 		.dport = dport,
 	};
 
-#ifdef ALLOW_ICMP_FRAG_NEEDED
-	/* When ALLOW_ICMP_FRAG_NEEDED is defined we allow all packets
-	 * of ICMP type 3 code 4 - Fragmentation Needed.
-	 */
-	if (proto == IPPROTO_ICMP) {
-		struct icmphdr icmphdr __align_stack_8;
-
-		if (ctx_load_bytes(ctx, off, &icmphdr, sizeof(icmphdr)) < 0)
-			return DROP_INVALID;
-
-		if (icmphdr.type == ICMP_DEST_UNREACH &&
-		    icmphdr.code == ICMP_FRAG_NEEDED) {
-			*proxy_port = 0;
-			return CTX_ACT_OK;
+#if defined(ALLOW_ICMP_FRAG_NEEDED) || defined(ENABLE_ICMP_RULE)
+	switch (ethertype) {
+	case ETH_P_IP:
+		if (proto == IPPROTO_ICMP) {
+			struct icmphdr icmphdr __align_stack_8;
+
+			if (ctx_load_bytes(ctx, off, &icmphdr, sizeof(icmphdr)) < 0)
+				return DROP_INVALID;
+
+# if defined(ALLOW_ICMP_FRAG_NEEDED)
+			if (icmphdr.type == ICMP_DEST_UNREACH &&
+			    icmphdr.code == ICMP_FRAG_NEEDED) {
+				*proxy_port = 0;
+				return CTX_ACT_OK;
+			}
+# endif
+
+# if defined(ENABLE_ICMP_RULE)
+			/* Convert from unsigned char to unsigned short
+			 * considering byte order(little-endian).
+			 * In the little-endian case, for example, 2byte data "AB"
+			 * convert to "BA".
+			 * Therefore, the "icmp_type" should be shifted not just casting.
+			 */
+			key.dport = (__u16)(icmphdr.type << 8);
```

Comment:
How can this be correct on all architectures without using some form of `ntohs()`?

Reply:
Hmm, looks like this was originally introduced by @chez-shanpu in https://github.com/cilium/cilium/pull/16516 .

Reply:
Yes, probably this line should use `bpf_htons()` instead of using shift.
I'll fix it and submit a PR.